File: ruby/arrays/arrays/passing_arrays_to_functions.rb
#!/usr/bin/ruby

def function_receiving_array_by_value(argument)
  puts("Function receiving array by value")
  puts("-- begin:")

  for i in 0..(argument.length() - 1)
    puts("before: argument[#{i}] = #{argument[i]}\n")
    puts("argument[#{i}] = argument[#{i}] * 2")

    argument[i] = argument[i] * 2

    puts("after: argument[#{i}] = #{argument[i]}\n")
  end

  puts("-- end.")

  argument
end

values = [9, 8, 7]

print("BEFORE: values = [ ")
values.each do |element|
  print("#{element} ")
end
puts("]")

result_values = function_receiving_array_by_value(values)

print("AFTER: values = [ ")
values.each do |element|
  print("#{element} ")
end
puts("]")
print("AFTER: result_values = [ ")
result_values.each do |element|
  print("#{element} ")
end
puts("]")

File: perl/arrays/arrays/passing_arrays_to_functions.pl
#!/usr/bin/perl

use strict;
use warnings;

sub function_receiving_array_by_reference {
  print("Function receiving array by value\n");
  print("-- begin:\n");

  my $argument_reference = $_[0];
  my @argument = @{$argument_reference};

  foreach my $key (0..(@argument - 1)) {
    print("before: \$argument[$key] = $argument[$key]\n");
    print("\$argument[$key] = argument[$key] * 2\n");

    $argument[$key] = $argument[$key] * 2;

    print("after: \$argument[$key] = $argument[$key]\n");
  }

  print("-- end.\n");

  return @argument;
}

my @values = (9, 8, 7);

print("BEFORE: \@values = [ ");
foreach (@values) {
  print $_ . " ";
}
print("]\n");

my @result_values = function_receiving_array_by_reference(\@values);

print("AFTER: \@values = [ ");
foreach (@values) {
  print $_ . " ";
}
print("]\n");
print("AFTER: \@result_values = [ ");
foreach (@result_values) {
  print $_ . " ";
}
print("]\n");
